File: 2381-shifting-letters-II/main.cpp
class Solution {
public:
    string shiftingLetters(string s, vector<vector<int>>& shifts) {
        int n = s.length();
        vector<int> diff(n+1, 0);

        for (const auto& shift: shifts) {
            int start = shift[0], end = shift[1], direction = shift[2];
            if (direction == 1) {
                diff[start]++;
                diff[end + 1]--;
            } else {
                diff[start]--;
                diff[end + 1]++;
            }
        }

        int currShift = 0;
        for (int i = 0; i < n; i++) {
            currShift = ((currShift + diff[i]) % 26 + 26) % 26;
            int curr = s[i] - 'a';
            int newChar = (curr + currShift) % 26;
            s[i] = newChar + 'a';
        }

        return s;
    }
};

File: 2381-shifting-letters-II/main.go
func shiftingLetters(s string, shifts [][]int) string {
    n := len(s)
    diff := make([]int, n+1)

    for _, shift := range shifts {
        start, end, direction := shift[0], shift[1], shift[2]
        if direction == 1 {
            diff[start]++
            diff[end+1]--
        }else {
            diff[start]--
            diff[end+1]++
        }
    }

    result := []rune(s)
    currShift := 0

    for i := 0; i < n; i++ {
        currShift = ((currShift + diff[i]) % 26 + 26) % 26
        curr := int(result[i] - 'a')
        newChar := (curr + currShift) % 26
        result[i] = rune(newChar + 'a')
    }

    return string(result)
}
